### PR TITLE
Insert logs into internal table

### PR DESCRIPTION
**Pull request by gavinyue**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [X] Feature
- [ ] Test and CI
- [ ] Code Refactoring

Design doc: https://doc.weixin.qq.com/doc/w3_AW0A-gb6AOInM1rKTLdS1G7q1Q6vl?scode=AJsA6gc3AA8X22kV0eAS0AvwapAOk

## Which issue(s) this PR fixes:

issue #   

- https://github.com/matrixorigin/matrixone/issues/8696
- https://github.com/matrixorigin/matrixone/issues/8702

## What this PR does / why we need it:

This will all logs/traces/metrics to insert into the internal tables. 

Still need to work on: 

1. Change the query strategy for other Accounts/Users to allow them to access statement_info and metric data. 
2. Add a config to switch off the SQL insert for BVT tests
3. More unit tests for SQLWriter

### CODE REVIEW

File: cmd/mo-service/main.go [hunk 1]
```diff
@@ -309,7 +309,7 @@ func initTraceMetric(ctx context.Context, st metadata.ServiceType, cfg *Config,
 	}
 
 	if !SV.DisableTrace || !SV.DisableMetric {
-		writerFactory = export.GetWriterFactory(fs, UUID, nodeRole, SV.LogsExtension)
+		writerFactory = export.GetWriterFactory(fs, UUID, nodeRole, table.CsvExtension)
```
Comment (xzxiong) [inline cmd/mo-service/main.go:312]: this should not be hard coded. should be control by config item as `SV.LogsExtension`
you could change default value to `csv`

Comment (gavinyue) [reply cmd/mo-service/main.go:312]: Will revert this after finishing the coding.

**Review by daviszhen — APPROVED** [review]

File: pkg/util/export/etl/csv.go [hunk 2]
```diff
@@ -58,13 +61,21 @@ func (w *CSVWriter) GetContent() string {
 }
 
 func (w *CSVWriter) FlushAndClose() (int, error) {
-	n, err := w.writer.WriteString(w.buf.String())
-	if err != nil {
-		return 0, err
+	// First write rows to sqlWriter, if not working, write to csv file
```
Comment (sukki37) [inline pkg/util/export/etl/csv.go:64]: Where is the csv file stored and what is the role of these csv files?

Comment (gavinyue) [reply pkg/util/export/etl/csv.go:64]: CSV files are stored in S3. They were designed for external tables. Now, they only act as a backup when can not insert into the internal table.

File: pkg/util/trace/impl/motrace/report_statement.go [hunk 3]
```diff
@@ -276,5 +277,10 @@ var ReportStatement = func(ctx context.Context, s *StatementInfo) error {
 	if !GetTracerProvider().IsEnable() {
 		return nil
 	}
+	// Filter out the MO_LOGGER SQL statements
+	if s.User == sqlWriter.MOLoggerUser {
```
Comment (xzxiong) [inline pkg/util/trace/impl/motrace/report_statement.go:281]: i believe cut off the suffix of statement text is better than drop the record.
may result like: `insert into system.statement_info valus (...)`

**Comment by xzxiong** [comment]
i think what is done:
- support insert logs/metrics/... data directly into ordinary table by sqlWriter
- could write csv file to S3 while sqlWriter not working

and the following feature should be continue:
- fetch all csv files data in S3, and insert into ordinary table with background task.
- config to switch off the ETLMerge task in `util/export/merge.go` 

other questions:
1. should we seperate diff metric item's data into diff ordinary table. for example: view system_metric.sql_statement_total ==> table system.sql_statement_total
2. as ordinary table, system.statement_info, system.rawlog, system_metrics.metric should have other index to improve query performance.

**Review by ouyuanning — APPROVED** [review]

**Review by aunjgr — APPROVED** [review]

**Review by aressu1985 — APPROVED** [review]

**Review by reusee — APPROVED** [review]

**Comment by xzxiong** [comment]
@gavinyue  sorry, i haven't seen any modification in ETL MergeTask with file `pkg/util/export/merge.go`.
as i konw, you would use merge task to read-s3-csv-and-reinsert-to-table ?

File: pkg/sql/plan/query_builder.go [hunk 4]
```diff
@@ -2063,6 +2064,22 @@ func (builder *QueryBuilder) buildTable(stmt tree.TableExpr, ctx *BindContext, p
 						return 0, err
 					}
 					builder.qry.Nodes[nodeID].FilterList = accountFilterExprs
+				} else if dbName == catalog.MO_SYSTEM_METRICS && tableName == catalog.MO_METRIC {
```
Comment (fengttt) [inline pkg/sql/plan/query_builder.go:2067]: This is almost for sure the wrong way to put in extra filter conditions.     There is nothing special about mo_metrics and mo_statement table.    We only have 3 special tables, cannot add more like this.

File: pkg/sql/util/util.go [hunk 5]
```diff
@@ -44,6 +44,15 @@ func SplitTableAndColumn(name string) (string, string) {
 	return schema, xs[len(xs)-1]
 }
 
+func TableIsLoggingTable(dbName string, tableName string) bool {
```
Comment (fengttt) [inline pkg/sql/util/util.go:47]: This is alslo wrong -- why did engine need to know these tables?   They are not special in any ways.

File: pkg/util/export/etl/sqlWriter/sqlWriter.go [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+}
+
+func (sw *BaseSqlWriter) GetContent() string {
+	return ""
+}
+func (sw *BaseSqlWriter) WriteRow(row *table.Row) error {
+	return nil
+}
+
+func generateInsertStatement(records [][]string, tbl *table.Table) (string, int, error) {
```
Comment (fengttt) [inline pkg/util/export/etl/sqlWriter/sqlWriter.go:56]: I am not sure this is the right way to do the insert.    I would rather do a prepared statement, then call insert as stmt.Execute(args...)

**Review by fengttt — COMMENTED** [review]
Where is the statement_info tabole created?    How about the other few tables like error_info?

**Comment by sukki37** [comment]
copilot:walkthrough

**Review by reusee — APPROVED** [review]

**Review by xzxiong — APPROVED** [review]

**Review by xzxiong — APPROVED** [review]

**Review by w-zr — APPROVED** [review]

**Review by aptend — APPROVED** [review]

**Review by LeftHandCold — APPROVED** [review]

**Review by m-schen — APPROVED** [review]

File: pkg/txn/rpc/server.go [hunk 7]
```diff
@@ -250,8 +250,7 @@ func (s *server) handleTxnRequest(ctx context.Context) {
 			if txnID, err := req.exec(); err != nil {
 				if s.rt.Logger().Enabled(zap.DebugLevel) {
 					s.rt.Logger().Error("handle txn request failed",
-						zap.String("txn-id", hex.EncodeToString(txnID)),
-						zap.Error(err))
+						zap.String("txn-id", hex.EncodeToString(txnID)))
```
Comment (zhangxu19830126) [inline pkg/txn/rpc/server.go:253]: why error field need removed?

**Review by zhangxu19830126 — CHANGES_REQUESTED** [review]

File: pkg/sql/colexec/connector/connector.go [hunk 8]
```diff
@@ -43,11 +43,11 @@ func Call(_ int, proc *process.Process, arg any, _ bool, _ bool) (bool, error) {
 	select {
 	case <-proc.Ctx.Done():
 		proc.PutBatch(bat)
-		logutil.Infof("proc context done during connector send")
+		logutil.Debugf("proc context done during connector send")
```
Comment (nnsgmsone) [inline pkg/sql/colexec/connector/connector.go:46]: This error is probably a real error, like a network disconnection or something. I think it needs to be info

File: pkg/sql/colexec/dispatch/dispatch.go [hunk 9]
```diff
@@ -121,7 +122,7 @@ func (arg *Argument) waitRemoteRegsReady(proc *process.Process) (bool, error) {
 			return false, moerr.NewInternalErrorNoCtx("wait notify message timeout")
 		case <-proc.Ctx.Done():
 			arg.ctr.prepared = true
-			logutil.Infof("conctx done during dispatch")
+			logutil.Debugf("conctx done during dispatch")
```
Comment (nnsgmsone) [inline pkg/sql/colexec/dispatch/dispatch.go:125]: This error is probably a real error, like a network disconnection or something. I think it needs to be info

File: pkg/util/trace/impl/motrace/report_statement.go (hunk 3 above)
Comment (gavinyue) [reply pkg/util/trace/impl/motrace/report_statement.go]: will review the statementinfo collected in the following steps

**Review by qingxinhome — APPROVED** [review]

File: pkg/sql/colexec/connector/connector.go (hunk 8 above)
Comment (gavinyue) [reply pkg/sql/colexec/connector/connector.go:46]: reverted

**Comment by gavinyue** [comment]
> @gavinyue sorry, i haven't seen any modification in ETL MergeTask with file `pkg/util/export/merge.go`. as i konw, you would use merge task to read-s3-csv-and-reinsert-to-table ?

Resolved as well

**Review by heni02 — APPROVED** [review]

File: pkg/objectio/writer.go [hunk 10]
```diff
@@ -434,7 +434,7 @@ func (w *objectWriterV1) AddBlock(blockMeta BlockObject, bat *batch.Batch, seqnu
 			rows = vec.Length()
 		} else if rows != vec.Length() {
 			warned++
-			logutil.Warnf("%s unmatched length, expect %d, get %d", bat.Attrs[i], rows, vec.Length())
+			logutil.Debugf("%s unmatched length, expect %d, get %d", bat.Attrs[i], rows, vec.Length())
```
Comment (XuPeng-SH) [inline pkg/objectio/writer.go:437]: low frequency, suggest keeping it

**Review by nnsgmsone — APPROVED** [review]

File: pkg/vm/engine/tae/catalog/catalog.go [hunk 11]
```diff
@@ -139,7 +140,7 @@ func (catalog *Catalog) InitSystemDB() {
 	}
 }
 func (catalog *Catalog) GCByTS(ctx context.Context, ts types.TS) {
-	logutil.Infof("GC Catalog %v", ts.ToString())
+	logutil.Debugf("GC Catalog %v", ts.ToString())
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/catalog/catalog.go:143]: low frequency, suggest to keep it as the original

File: pkg/vm/engine/tae/db/checkpoint/testutils.go [hunk 12]
```diff
@@ -156,7 +156,7 @@ func (r *runner) ForceIncrementalCheckpoint(end types.TS) error {
 		return err
 	}
 	entry.SetState(ST_Finished)
-	logutil.Infof("%s is done, takes %s", entry.String(), time.Since(now))
+	logutil.Debugf("%s is done, takes %s", entry.String(), time.Since(now))
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/db/checkpoint/testutils.go:159]: low frequency, suggest keeping it as the original

File: pkg/vm/engine/tae/db/scannerop.go [hunk 13]
```diff
@@ -169,7 +169,7 @@ func (d *deletableSegBuilder) finish() []*catalog.SegmentEntry {
 	copy(ret[:len(d.segCandids)], d.segCandids)
 	copy(ret[len(d.segCandids):], d.nsegCandids)
 	if cnt := len(d.nsegCandids); cnt != 0 {
-		logutil.Info("Mergeblocks deletable nseg", zap.Int("cnt", cnt))
+		logutil.Debug("Mergeblocks deletable nseg", zap.Int("cnt", cnt))
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/db/scannerop.go:172]: low frequency, suggest keeping it as the original

File: pkg/vm/engine/tae/db/scannerop.go [hunk 14]
```diff
@@ -207,7 +207,7 @@ func (ml *mergeLimiter) canMerge(tid uint64, totalRow int, blks int) bool {
 		return false
 	}
 	if totalRow > constMergeRightNow {
-		logutil.Infof("Mergeblocks %d merge right now: %d rows %d blks", tid, totalRow, blks)
+		logutil.Debugf("Mergeblocks %d merge right now: %d rows %d blks", tid, totalRow, blks)
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/db/scannerop.go:210]: low frequency, suggest keeping it as the original

File: pkg/vm/engine/tae/db/scannerop.go [hunk 15]
```diff
@@ -225,7 +225,7 @@ func (ml *mergeLimiter) canMerge(tid uint64, totalRow int, blks int) bool {
 		// a lot of things happened in the past scan interval...
 		st.ttl = ml.ttl(totalRow)
 		st.lastTotalRow = totalRow
-		logutil.Infof("Mergeblocks delta %d on table %d, resched to %v", d, tid, st.ttl)
+		logutil.Debugf("Mergeblocks delta %d on table %d, resched to %v", d, tid, st.ttl)
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/db/scannerop.go:228]: low frequency, suggest keeping it as the original

File: pkg/vm/engine/tae/db/scannerop.go [hunk 16]
```diff
@@ -322,7 +322,7 @@ func (s *MergeTaskBuilder) trySchedMergeTask() {
 			logutil.Infof("[Mergeblocks] Schedule del seg errinfo=%v", err)
 			return
 		}
-		logutil.Infof("[Mergeblocks] Scheduled | del %d seg", len(mergedSegs))
+		logutil.Debugf("[Mergeblocks] Scheduled | del %d seg", len(mergedSegs))
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/db/scannerop.go:325]: low frequency, suggest keeping it as the original

File: pkg/vm/engine/tae/db/scannerop.go [hunk 17]
```diff
@@ -345,7 +345,7 @@ func (s *MergeTaskBuilder) trySchedMergeTask() {
 			s.limiter.IncActiveCount()
 			task.AddObserver(s.limiter)
 		}
-		logutil.Infof("[Mergeblocks] Scheduled | Scopes=[%d],[%d]%s",
+		logutil.Debugf("[Mergeblocks] Scheduled | Scopes=[%d],[%d]%s",
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/db/scannerop.go:348]: low frequency, suggest keeping it as the original

File: pkg/vm/engine/tae/db/scannerop.go [hunk 18]
```diff
@@ -367,7 +367,7 @@ func (s *MergeTaskBuilder) PreExecute() error {
 
 	// print stats for every 50s (default)
 	if s.runCnt%10 == 0 {
-		logutil.Infof("Mergeblocks stats: %s", s.limiter.String())
+		logutil.Debugf("Mergeblocks stats: %s", s.limiter.String())
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/db/scannerop.go:370]: low frequency, suggest keeping it as the original

File: pkg/vm/engine/tae/db/scannerop.go [hunk 19]
```diff
@@ -376,7 +376,7 @@ func (s *MergeTaskBuilder) PreExecute() error {
 			logutil.Infof("Mergeblocks available mem: %dg", stats.Available/(1<<30))
 			if limit := int32(stats.Available / const4GBytes); limit != s.limiter.concurrentMergeLimit && limit > 1 {
 				s.limiter.concurrentMergeLimit = limit
-				logutil.Infof("Mergeblocks set concurrency limit %d", limit)
+				logutil.Debugf("Mergeblocks set concurrency limit %d", limit)
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/db/scannerop.go:379]: low frequency, suggest keeping it as the original

File: pkg/vm/engine/tae/logtail/mgr.go [hunk 20]
```diff
@@ -186,7 +186,7 @@ func (mgr *Manager) GCByTS(ctx context.Context, ts types.TS) {
 	}
 	mgr.truncated = ts
 	cnt := mgr.table.TruncateByTimeStamp(ts)
-	logutil.Info("[logtail] GC", zap.String("ts", ts.ToString()), zap.Int("deleted", cnt))
+	logutil.Debug("[logtail] GC", zap.String("ts", ts.ToString()), zap.Int("deleted", cnt))
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/logtail/mgr.go:189]: low frequency, suggest keeping it as the original

File: pkg/vm/engine/tae/logtail/service/server.go [hunk 21]
```diff
@@ -386,7 +386,7 @@ func (s *LogtailServer) logtailSender(ctx context.Context) {
 				return
 			}
 
-			logger.Info("handle subscription asynchronously", zap.Any("table", sub.req.Table))
+			logger.Debug("handle subscription asynchronously", zap.Any("table", sub.req.Table))
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/logtail/service/server.go:389]: low frequency, suggest keeping it as the original

File: pkg/vm/engine/tae/logtail/service/session.go [hunk 22]
```diff
@@ -219,7 +219,7 @@ func NewSession(
 		heartbeatTimer:    time.NewTimer(heartbeatInterval),
 	}
 
-	ss.logger.Info("initialize new session for morpc stream")
+	ss.logger.Debug("initialize new session for morpc stream")
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/logtail/service/session.go:222]: low frequency, suggest keeping it as the original

File: pkg/vm/engine/tae/logtail/service/session.go [hunk 23]
```diff
@@ -232,7 +232,7 @@ func NewSession(
 
 			case msg, ok := <-ss.sendChan:
 				if !ok {
-					ss.logger.Info("session sender channel closed")
+					ss.logger.Debug("session sender channel closed")
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/logtail/service/session.go:235]: low frequency, suggest keeping it as the original

File: pkg/vm/engine/tae/logtail/service/session.go [hunk 24]
```diff
@@ -269,7 +269,7 @@ func NewSession(
 
 // Drop closes sender goroutine.
 func (ss *Session) PostClean() {
-	ss.logger.Info("clean session for morpc stream")
+	ss.logger.Debug("clean session for morpc stream")
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/logtail/service/session.go:272]: low frequency, suggest keeping it as the original

File: pkg/vm/engine/tae/logtail/service/session.go [hunk 25]
```diff
@@ -331,7 +331,7 @@ func (ss *Session) FilterLogtail(tails ...wrapLogtail) []logtail.TableLogtail {
 		if state, ok := ss.tables[t.id]; ok && state == TableSubscribed {
 			qualified = append(qualified, t.tail)
 		} else {
-			ss.logger.Info("table not subscribed, filter out",
+			ss.logger.Debug("table not subscribed, filter out",
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/logtail/service/session.go:334]: low frequency, suggest keeping it as the original

File: pkg/vm/engine/tae/logtail/service/session.go [hunk 26]
```diff
@@ -404,7 +404,7 @@ func (ss *Session) SendErrorResponse(
 func (ss *Session) SendSubscriptionResponse(
 	sendCtx context.Context, tail logtail.TableLogtail,
 ) error {
-	ss.logger.Info("send subscription response", zap.Any("table", tail.Table), zap.String("To", tail.Ts.String()))
+	ss.logger.Debug("send subscription response", zap.Any("table", tail.Table), zap.String("To", tail.Ts.String()))
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/logtail/service/session.go:407]: low frequency, suggest keeping it as the original. subsciption is only frequent in BVT scenario

File: pkg/vm/engine/tae/logtail/service/session.go [hunk 27]
```diff
@@ -419,7 +419,7 @@ func (ss *Session) SendSubscriptionResponse(
 func (ss *Session) SendUnsubscriptionResponse(
 	sendCtx context.Context, table api.TableID,
 ) error {
-	ss.logger.Info("send unsubscription response", zap.Any("table", table))
+	ss.logger.Debug("send unsubscription response", zap.Any("table", table))
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/logtail/service/session.go:422]: low frequency, suggest keeping it as the original

File: pkg/vm/engine/tae/logtail/table.go [hunk 28]
```diff
@@ -197,7 +197,7 @@ func (table *TxnTable) ForeachRowInBetween(
 			minTs = blk.bornTS
 			return false
 		})
-		logutil.Warn("[logtail] fetch with too small ts", zap.String("ts", from.ToString()), zap.String("minTs", minTs.ToString()))
+		logutil.Debug("[logtail] fetch with too small ts", zap.String("ts", from.ToString()), zap.String("minTs", minTs.ToString()))
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/logtail/table.go:200]: low frequency, suggest keeping it as the original

File: pkg/vm/engine/tae/rpc/handle.go [hunk 29]
```diff
@@ -957,21 +962,21 @@ func openTAE(targetDir string, opt *options.Options) (tae *db.DB, err error) {
 		mask := syscall.Umask(0)
 		if err := os.MkdirAll(targetDir, os.FileMode(0755)); err != nil {
 			syscall.Umask(mask)
-			logutil.Infof("Recreate dir error:%v\n", err)
+			logutil.Debugf("Recreate dir error:%v", err)
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/rpc/handle.go:965]: very very low frequency, suggest keeping it as the original

File: pkg/vm/engine/tae/tables/jobs/compactblk.go [hunk 30]
```diff
@@ -143,7 +144,7 @@ func (task *compactBlockTask) Name() string {
 }
 
 func (task *compactBlockTask) Execute(ctx context.Context) (err error) {
-	logutil.Info("[Start]", common.OperationField(task.Name()),
+	logutil.Debug("[Start]", common.OperationField(task.Name()),
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/tables/jobs/compactblk.go:147]: low frequency, suggest keeping it as the original

File: pkg/vm/engine/tae/tables/jobs/compactblk.go [hunk 31]
```diff
@@ -184,7 +185,7 @@ func (task *compactBlockTask) Execute(ctx context.Context) (err error) {
 				// we can't create appendable seg here because compaction can be rollbacked.
 				// so just wait until the new appendable seg is available.
 				// actually this log can barely be printed.
-				logutil.Infof("do not compact on seg %s %d, wait", curSeg.ID.ToString(), curSeg.GetNextObjectIndex())
+				logutil.Debugf("do not compact on seg %s %d, wait", curSeg.ID.ToString(), curSeg.GetNextObjectIndex())
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/tables/jobs/compactblk.go:188]: low frequency, suggest keeping it as the original

File: pkg/vm/engine/tae/tables/jobs/compactblk.go [hunk 32]
```diff
@@ -286,7 +287,7 @@ func (task *compactBlockTask) Execute(ctx context.Context) (err error) {
 	if task.created != nil {
 		createdStr = task.created.Fingerprint().BlockString()
 	}
-	logutil.Info("[Done]",
+	logutil.Debug("[Done]",
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/tables/jobs/compactblk.go:290]: low frequency, suggest keeping it as the original

**Review by zhangxu19830126 — APPROVED** [review]

**Review by XuPeng-SH — APPROVED** [review]

**Review by sukki37 — APPROVED** [review]

**Review by sukki37 — APPROVED** [review]

**Pull request merged**